syntax = "proto3";
package wodson;

message DataMatrix {

  message Column {

    enum AggregateEnum {
      NONE = 0;
      COUNT = 1;
      DCOUNT = 2;
      MIN = 3;
      MAX = 4;
      AVG = 5;
      STDDEV = 6;
      SUM = 7;
      DISTINCT = 8;
      POINT = 9;
    }

    enum DatatypeEnum {
      DT_UNKNOWN = 0;
      DT_STRING = 1;
      DT_SHORT = 2;
      DT_FLOAT = 3;
      DT_BOOLEAN = 4;
      DT_BYTE = 5;
      DT_LONG = 6;
      DT_DOUBLE = 7;
      DT_LONGLONG = 8;
      DT_ID = 9;
      DT_DATE = 10;
      DT_BYTESTR = 11;
      DT_COMPLEX = 12;
      DT_DCOMPLEX = 13;
      DS_STRING = 14;
      DS_SHORT = 15;
      DS_FLOAT = 16;
      DS_BOOLEAN = 17;
      DS_BYTE = 18;
      DS_LONG = 19;
      DS_DOUBLE = 20;
      DS_LONGLONG = 21;
      DS_COMPLEX = 22;
      DS_DCOMPLEX = 23;
      DS_ID = 24; 
      DS_DATE = 25;
      DS_BYTESTR = 26;
      DT_EXTERNALREFERENCE = 27;
      DS_EXTERNALREFERENCE = 28;
      DT_ENUM = 29;
      DS_ENUM = 30;
    }
      
    message StringArray { repeated string values = 1; };
    message LongArray { repeated int32 values = 1; };
    message FloatArray { repeated float values = 1; };
    message BooleanArray { repeated bool values = 1; };
    message ByteArray { bytes values = 1; };
    message DoubleArray { repeated double values = 1; };
    message LongLongArray { repeated int64 values = 1; };
    message BytestrArray { repeated bytes values = 1; };
    message UnknownArray { 
      DatatypeEnum datatype = 1;
      
      StringArray   dtString = 2; // DT_STRING, DT_DATE, DT_EXTERNALREFERENCE
      LongArray     dtLong = 3; // DT_SHORT, DT_LONG, DT_ENUM
      FloatArray    dtFloat = 4; // DT_FLOAT, DT_COMPLEX
      BooleanArray  dtBoolean = 5; // DT_BOOLEAN
      ByteArray     dtByte = 6; // DT_BYTE
      DoubleArray   dtDouble = 7; // DT_DOUBLE, DT_DCOMPLEX
      LongLongArray dtLonglong = 8; // DT_LONGLONG, DT_ID
      BytestrArray  dtBytestr = 9; // DT_BYTESTR
    };

    string name = 1;
    string baseName = 2;
    AggregateEnum aggregate = 3;
    DatatypeEnum datatype = 4;
    repeated bool flags = 5;

    StringArray   dtString = 6; // DT_STRING, DT_DATE, DT_EXTERNALREFERENCE
    LongArray     dtLong = 7; // DT_SHORT, DT_LONG, DT_ENUM
    FloatArray    dtFloat = 8; // DT_FLOAT, DT_COMPLEX
    BooleanArray  dtBoolean = 9; // DT_BOOLEAN
    ByteArray     dtByte = 10; // DT_BYTE
    DoubleArray   dtDouble = 11; // DT_DOUBLE, DT_DCOMPLEX
    LongLongArray dtLonglong = 12; // DT_LONGLONG, DT_ID
    BytestrArray  dtBytestr = 13; // DT_BYTESTR

    repeated StringArray    dsString = 14; // DS_STRING, DS_DATE, DS_EXTERNALREFERENCE
    repeated LongArray      dsLong = 15; // DS_SHORT, DS_LONG, DS_ENUM
    repeated FloatArray     dsFloat = 16; // DS_FLOAT, DS_COMPLEX
    repeated BooleanArray   dsBoolean = 17; // DS_BOOLEAN
    repeated ByteArray      dsByte = 18; // DS_BYTE
    repeated DoubleArray    dsDouble = 19; // DS_DOUBLE, DS_DCOMPLEX
    repeated LongLongArray  dsLonglong = 20; // DS_LONGLONG, DS_ID
    repeated BytestrArray   dsBytestr = 21; // DS_BYTESTR
      
    repeated UnknownArray dtUnknown = 22; // DT_UNKNOW
  }

  string name = 1;
  string baseName = 2;
  int64 rowSkipCount = 3;
  int64 seqSkipCount = 4;
  repeated Column columns = 5;
}